internal/blockchain/ethereum/ethereum.go

```diff
@@ -956,121 +981,167 @@ func (e *Ethereum) GenerateFFI(ctx context.Context, generationRequest *core.FFIG
-			Type:         arg.Type,
-			InternalType: arg.InternalType,
-			Indexed:      arg.Indexed,
-		},
+	for i, output := range abiFunction.Outputs {
+		typeComponent, err := output.TypeComponentTreeCtx(ctx)
+		if err != nil {
+			return nil, err
+		}
+		schema := e.getSchemaForABIInput(ctx, typeComponent)
+		param := &core.FFIParam{
+			Name:   output.Name,
+			Schema: fftypes.JSONAnyPtr(schema.ToJSON()),
+		}
+		returns[i] = param
 	}
-	var properties map[string]*Schema
-	if len(arg.Components) > 0 {
-		properties = e.getSchemaForObjectComponents(arg)
+	if abiFunction.StateMutability != "" {
+		details["stateMutability"] = string(abiFunction.StateMutability)
 	}
-	if s.Type == arrayType {
-		levels := strings.Count(arg.Type, "[]")
-		innerType := e.getFFIType(strings.ReplaceAll(arg.Type, "[]", ""))
-		innerSchema := &Schema{
-			Type: innerType,
-		}
-		if len(arg.Components) > 0 {
-			innerSchema.Properties = e.getSchemaForObjectComponents(arg)
+	if abiFunction.Payable {
+		details["payable"] = true
+	}
+	if abiFunction.Constant {
+		details["constant"] = true
+	}
+	return &core.FFIMethod{
+		Name:    abiFunction.Name,
+		Params:  params,
+		Returns: returns,
+		Details: details,
+	}, nil
+}
+
+func (e *Ethereum) convertABIEventToFFIEvent(ctx context.Context, abiEvent *abi.Entry) (*core.FFIEvent, error) {
+	params := make([]*core.FFIParam, len(abiEvent.Inputs))
+	details := map[string]interface{}{}
+	for i, output := range abiEvent.Inputs {
+		typeComponent, err := output.TypeComponentTreeCtx(ctx)
+		if err != nil {
+			return nil, err
 		}
-		for i := 1; i < levels; i++ {
-			innerSchema = &Schema{
-				Type:  arrayType,
-				Items: innerSchema,
-			}
+		schema := e.getSchemaForABIInput(ctx, typeComponent)
+		param := &core.FFIParam{
+			Name:   output.Name,
+			Schema: fftypes.JSONAnyPtr(schema.ToJSON()),
 		}
-		s.Items = innerSchema
-	} else {
-		s.Properties = properties
+		params[i] = param
 	}
-	return s
+	if abiEvent.Anonymous {
+		details["anonymous"] = true
+	}
+	return &core.FFIEvent{
+		FFIEventDefinition: core.FFIEventDefinition{
+			Name:    abiEvent.Name,
+			Params:  params,
+			Details: details,
+		},
+	}, nil
 }
 
-func (e *Ethereum) getSchemaForObjectComponents(arg ABIArgumentMarshaling) map[string]*Schema {
-	m := make(map[string]*Schema, len(arg.Components))
-	for i, component := range arg.Components {
-		componentSchema := e.getSchema(component)
-		componentSchema.Details.Index = new(int)
-		*componentSchema.Details.Index = i
-		m[component.Name] = componentSchema
+func (e *Ethereum) getSchemaForABIInput(ctx context.Context, typeComponent abi.TypeComponent) *Schema {
+	schema := &Schema{
+		Details: &paramDetails{
+			Type:         typeComponent.Parameter().Type,
+			InternalType: typeComponent.Parameter().InternalType,
+			Indexed:      typeComponent.Parameter().Indexed,
+		},
+	}
+	switch typeComponent.ComponentType() {
+	case abi.ElementaryComponent:
+		t := e.getFFIType(typeComponent.ElementaryType().String())
```

Comment:
I think we still need this bit of code, because somewhere we need a mapping from ABI type (e.g. `uint256`) to the JSON input type we will accept on the API (e.g. `integer`)